File: 프로그래머스/lv3/43162. 네트워크/네트워크.java
class Solution { //dfs
    static int[][] g_computers;
    static int g_n;
    int t_row, t_col;
    
    public int solution(int n, int[][] computers) {
        g_computers=computers;
        g_n = n;
        t_row = computers.length;
        t_col = computers[0].length;
        
        int result = 0;
        
        for(int r=0; r<t_row; r++){
            for(int c=0; c<t_col; c++){

                if(g_computers[r][c]==1){
                    result++;
                    g_computers[r][c]=0;
                    g_computers[c][r]=0;
                    dfs(c);

                }
            }
        }
        
        return result;
    }
    
    void dfs(int row){
    System.out.println(row);
       for(int col =0; col<t_col; col++){
           if(g_computers[row][col]==1){
               g_computers[row][col] = 0;
               g_computers[col][row] = 0;
                dfs(col);
           }
       }
    }
}

File: 프로그래머스/lv3/43162. 네트워크/네트워크.py
def solution(n, computers):
    if n == 1:
        return 1
    if n == 2:
        if computers[0][1] == 1:
            return 1
        else: return 2
    
    
    global parent
    parent = [i for i in range(n)]
    for row in range(n):
        for col in range(n):
            if computers[row][col] == 1:
                union(row,col)
    for i in range(n): #이부분이 없으면 TC9번을 틀린다.
        findParent(i)
    return len(set(parent))

def findParent(x):
    if x != parent[x]:
        parent[x] = findParent(parent[x])
    return parent[x]
    
def union(a,b):
    a = findParent(a)
    b = findParent(b)
    
    if a<b:
        parent[b] = a
    else:
        parent[a] = b